File: sprint03/KDH/PGS_42888.py
# 오픈채팅방
from collections import defaultdict


def solution(record):
    answer = []
    name_dic = defaultdict(list)
    action_dic = {'Enter': '들어왔습니다.', 'Leave': '나갔습니다.'}
    res = []

    for i in range(len(record)):
        record[i] = record[i].split(" ")
        if record[i][0] == "Enter":
            res.append(["Enter", record[i][1]])
            name_dic[record[i][1]] = record[i][2]
        elif record[i][0] == "Leave":
            res.append(["Leave", record[i][1]])
        else:
            name_dic[record[i][1]] = record[i][2]

    for i in range(len(res)):
        answer.append(name_dic[res[i][1]]+"님이 " + action_dic[res[i][0]])

    return answer

File: sprint03/JMS/PGS_42888.cpp
#include <string>
#include <vector>
#include <map>
#include <sstream>

using namespace std;

map<string, string> idMap;

vector<string> solution(vector<string> record) {
    vector<string> answer, tempanswer;
    
    for(auto p : record){
        stringstream ss(p);
        ss.str(p);
        string info[3];
        int i=0, infoindex=0;
        
        while(ss >> info[infoindex++]);
        
        if(info[0].compare("Enter") == 0){
            idMap[info[1]] = info[2];
            tempanswer.push_back("E" + info[1]);
        }
        else if(info[0].compare("Leave") == 0)
            tempanswer.push_back("L" + info[1]);
        else if(info[0].compare("Change") == 0)
            idMap[info[1]] = info[2];
    }
    
    for(auto p : tempanswer){
        if(p[0] == 'E'){
            string nickname  = idMap[p.substr(1)];
            answer.push_back(nickname.append("님이 들어왔습니다."));
        }
        else if(p[0] == 'L'){
            string nickname  = idMap[p.substr(1)];
            answer.push_back(nickname.append("님이 나갔습니다."));
        }
    }
    
    return answer;
}
